src/os/bluestore/BlueStore.cc

```diff
@@ -6164,10 +6168,17 @@ void BlueStore::_txc_state_proc(TransContext *txc)
       }
       txc->shared_blobs_written.clear();
       if (g_conf->bluestore_sync_submit_transaction) {
-	_txc_finalize_kv(txc, txc->t);
-	txc->kv_submitted = true;
-	int r = db->submit_transaction(txc->t);
-	assert(r == 0);
+	if (txc->last_nid >= nid_max ||
+	    txc->last_blobid >= blobid_max) {
+	  dout(20) << __func__
+		   << " last_{nid,blobid} exceeds max, submit via kv thread"
+		   << dendl;
+	} else {
+	  _txc_finalize_kv(txc, txc->t);
+	  txc->kv_submitted = true;
+	  int r = db->submit_transaction(txc->t);
+	  assert(r == 0);
+	}
       }
```

Comment:
@liewegas .  Consider this:
txc1/txc2 last_nid both are larger nid_max.
  thread1                           thread2
 update nid_max
                                        submit_transaction
 submit_transaction but failed.
When replay check, txc2->last_nid > nid_max.
I think we in kv_sync_thread, we must make sure the new nid_max must update and then update nid_max.